File: src/checks/playback/opaque.cpp
#include <QTest>
#include <vector>

#include "checks/playback/tst_xcmd.h"

namespace checks {

// A write past an unknown-selector epoch keeps the opaque bytes entirely
// untouched (no removal, no re-role).
void XcmdTest::writeAfterOpaqueEpochLeavesItUntouched()
{
    const auto events = makeEvents(
        ev(0, 1, 0, xcmd::kSelectorController, 0x03), ev(1, 2, 0, xcmd::kPayloadController, 9),
        ev(2, 5, 0, xcmd::kSelectorController, 0x08), ev(3, 6, 0, xcmd::kPayloadController, 34));
    const std::vector<xcmd::PointWrite> writes = {{8, xcmd::kEchoVolumeLane, 40, 0, 0}};
    const auto patch = xcmd::rewritePoints(events, {}, writes);
    QVERIFY2(patch && patch->removeEvents.empty() && patch->inserts.size() == 2 &&
                 patch->inserts[0].tick == 8 && patch->inserts[1].tick == 8,
             "write after an opaque epoch touched the opaque traffic");
}

// A write at an opaque epoch's interior tick would re-roll surviving bytes:
// rejected, including the opaque selector's own tick.
void XcmdTest::writeInsideOpaqueEpochRejected()
{
    const auto events = makeEvents(ev(0, 1, 0, xcmd::kSelectorController, 0x03),
                                   ev(1, 2, 0, xcmd::kPayloadController, 9),
                                   ev(2, 3, 0, xcmd::kPayloadController, 10));
    std::vector<xcmd::PointWrite> writes = {{2, xcmd::kEchoVolumeLane, 40, 0, 0}};
    QVERIFY2(!xcmd::rewritePoints(events, {}, writes),
             "write inside an opaque epoch was not rejected");
    writes[0].tick = 1; // the opaque selector's own tick
    QVERIFY2(!xcmd::rewritePoints(events, {}, writes),
             "write on an opaque epoch's first tick was not rejected");
}

// A write inside a leading stray run's occupied span is rejected too.
void XcmdTest::writeInsideStrayRunSpanRejected()
{
    const auto events = makeEvents(ev(0, 1, 0, xcmd::kPayloadController, 99),
                                   ev(1, 2, 0, xcmd::kPayloadController, 98));
    const std::vector<xcmd::PointWrite> writes = {{1, xcmd::kEchoVolumeLane, 40, 0, 0}};
    QVERIFY2(!xcmd::rewritePoints(events, {}, writes),
             "write inside a stray-run span was not rejected");
}

// Remove identities that are not known points reject: opaque member,
// selector byte, and unknown index.
void XcmdTest::unknownRemoveIdentitiesRejected()
{
    const auto events = makeEvents(
        ev(0, 1, 0, xcmd::kSelectorController, 0x03), ev(1, 2, 0, xcmd::kPayloadController, 9),
        ev(2, 3, 0, xcmd::kSelectorController, 0x08), ev(3, 4, 0, xcmd::kPayloadController, 34));
    const std::vector<uint64_t> opaqueMember = {1};
    QVERIFY2(!xcmd::rewritePoints(events, opaqueMember, {}),
             "removing an opaque member was not rejected");
    const std::vector<uint64_t> selectorByte = {2};
    QVERIFY2(!xcmd::rewritePoints(events, selectorByte, {}),
             "removing a known selector was not rejected");
    const std::vector<uint64_t> staleIdentity = {999};
    QVERIFY2(!xcmd::rewritePoints(events, staleIdentity, {}),
             "removing a stale identity was not rejected");
}

} // namespace checks

File: src/checks/playback/opaque.swift
import Foundation
import PorydawCore

internal func xcmdPairedStrayWrite(_ report: CheckReport) {
    let stray = [
        Xcmd.Event(index: 0, tick: 4, stream: 0, controller: 0x1D, value: 7),
    ]
    report.expect(Xcmd.rewrite(stray, removing: [], writing: [
        Xcmd.PointWrite(tick: 4, lane: 0xFB, value: 20, stream: 0, channel: 0),
    ]) == nil, cppID: "xcmdcheck/XcmdTest::writeInsideStrayRunSpanRejected",
    message: "write inside a stray payload run rejects")
}

private func xcmdPairWriteAfterOpaqueEpochLeavesItUntouched(_ report: CheckReport) {
    let events = [
        xcmdPairEvent(0, 1, 0, Xcmd.selectorController, 0x03), xcmdPairEvent(1, 2, 0, Xcmd.payloadController, 9),
        xcmdPairEvent(2, 5, 0, Xcmd.selectorController, 0x08), xcmdPairEvent(3, 6, 0, Xcmd.payloadController, 34)]
    let writes: [Xcmd.PointWrite] = [Xcmd.PointWrite(tick: 8, lane: Xcmd.echoVolumeLane, value: 40, stream: 0, channel: 0)]
    let patch = Xcmd.rewrite(events, removing: [], writing: writes)
    report.expect(patch != nil && patch!.removeEvents.isEmpty && patch!.inserts.count == 2 &&
                 patch!.inserts[0].tick == 8 && patch!.inserts[1].tick == 8,
        cppID: "xcmdcheck/XcmdTest::writeAfterOpaqueEpochLeavesItUntouched",
        message: "write after an opaque epoch touched the opaque traffic")
}

private func xcmdPairWriteInsideOpaqueEpochRejected(_ report: CheckReport) {
    let events = [xcmdPairEvent(0, 1, 0, Xcmd.selectorController, 0x03),
                                   xcmdPairEvent(1, 2, 0, Xcmd.payloadController, 9),
                                   xcmdPairEvent(2, 3, 0, Xcmd.payloadController, 10)]
    var writes: [Xcmd.PointWrite] = [Xcmd.PointWrite(tick: 2, lane: Xcmd.echoVolumeLane, value: 40, stream: 0, channel: 0)]
    report.expect(Xcmd.rewrite(events, removing: [], writing: writes) == nil,
        cppID: "xcmdcheck/XcmdTest::writeInsideOpaqueEpochRejected",
        message: "write inside an opaque epoch was not rejected")
    writes[0].tick = 1
    report.expect(Xcmd.rewrite(events, removing: [], writing: writes) == nil,
        cppID: "xcmdcheck/XcmdTest::writeInsideOpaqueEpochRejected",
        message: "write on an opaque epoch's first tick was not rejected")
}

private func xcmdPairWriteInsideStrayRunSpanRejected(_ report: CheckReport) {
    let events = [xcmdPairEvent(0, 1, 0, Xcmd.payloadController, 99),
                                   xcmdPairEvent(1, 2, 0, Xcmd.payloadController, 98)]
    let writes: [Xcmd.PointWrite] = [Xcmd.PointWrite(tick: 1, lane: Xcmd.echoVolumeLane, value: 40, stream: 0, channel: 0)]
    report.expect(Xcmd.rewrite(events, removing: [], writing: writes) == nil,
        cppID: "xcmdcheck/XcmdTest::writeInsideStrayRunSpanRejected",
        message: "write inside a stray-run span was not rejected")
}

private func xcmdPairUnknownRemoveIdentitiesRejected(_ report: CheckReport) {
    let events = [
        xcmdPairEvent(0, 1, 0, Xcmd.selectorController, 0x03), xcmdPairEvent(1, 2, 0, Xcmd.payloadController, 9),
        xcmdPairEvent(2, 3, 0, Xcmd.selectorController, 0x08), xcmdPairEvent(3, 4, 0, Xcmd.payloadController, 34)]
    let opaqueMember: [UInt64] = [1]
    report.expect(Xcmd.rewrite(events, removing: opaqueMember, writing: []) == nil,
        cppID: "xcmdcheck/XcmdTest::unknownRemoveIdentitiesRejected",
        message: "removing an opaque member was not rejected")
    let selectorByte: [UInt64] = [2]
    report.expect(Xcmd.rewrite(events, removing: selectorByte, writing: []) == nil,
        cppID: "xcmdcheck/XcmdTest::unknownRemoveIdentitiesRejected",
        message: "removing a known selector was not rejected")
    let staleIdentity: [UInt64] = [999]
    report.expect(Xcmd.rewrite(events, removing: staleIdentity, writing: []) == nil,
        cppID: "xcmdcheck/XcmdTest::unknownRemoveIdentitiesRejected",
        message: "removing a stale identity was not rejected")
}

internal func runXcmdOpaqueOriginalChecks(_ report: CheckReport) {
    xcmdPairWriteAfterOpaqueEpochLeavesItUntouched(report)
    xcmdPairWriteInsideOpaqueEpochRejected(report)
    xcmdPairWriteInsideStrayRunSpanRejected(report)
    xcmdPairUnknownRemoveIdentitiesRejected(report)
}
